### ISSUE TITLE
[Request] Add option to use external function in order to read the API JSON response

### ISSUE THREAD

**Issue by justa9** [stephematician/statsnbaR#1]
Hello!

First of all, thanks for the package. I do find this package superior to the alternative mentioned in the README, since the latter requires a whole lot of spurious packages who either have been removed from CRAN or do not compile under all OS's, thus breaking portability.

The thing is, nowadays, API access through terminal-based methods (curl, etc) consistently timeouts for most (if not all) users. So, I would suggest adding an option to read a JSON file that holds the API response (or giving the user the ability to specify a function that can access the API).

Thanks you.

**Comment by stephematician** [comment]
Hi, thanks for your interest!

If a user is already able to download the data they need into JSON format, I'm not sure if there is much additional value added by my package in it's current form - this package is not much beyond a wrapper for the HTTP requests for the API endpoint of NBA.com.

What sort of user supplied functions were you thinking of?

**Comment by justa9** [comment]
Allow me to disagree on the usefulness of your package: it has a decent modular structure and seems to adhere to certain coding guidelines. This make debugging/extending easier.

The user supplied function can be anything that can connect to the API and return back a JSON object, from a system call (using ```wget``` and according options), to the forking of a headless browser (currently not blocked by the API), to a simple wrapper around ```base::open``` (for an already downloaded JSON file).

Thus, if the user where to pass a function as an argument to a new argument of ```api_scrape``` then instead of running the internal code, the user-supplied function would be executed and return the JSON text.

Do the above make any sense?


PS: Currently, ```team_game_logs``` reports a warning concerning a mapping procedure for ```nc[k]``` ("number of items to replace is not a multiple of replacement length") . Should I open a new issue for this, or just post a simple writeup here?

**Comment by stephematician** [comment]
Thanks, I'll have a look into the issue as well, no need to write up.

**Comment by stephematician** [comment]
1. I added an argument called `method` (terrible name, but I'll change this soon once we get the mechanics of it down-pat) that can be passed on to the current 'suite' of functions that are exported by statsnbaR such as `get_players` and so on. If `method` exists and is a function, it is called instead of `GET` during the retrieval of the JSON object. To start with, the `url` and `headers` arguments that would normally be passed to the call to `GET` are passed to the call to `method`. Basically, the `method` function must return the same data as though `content()` was called on the response object that would be returned by `GET`. If it doesn't, then the work that `statsnbaR` does to tidy up the types/values of the returned data may fail.

  2. All the functions that collect the data now accept variable (or additional) arguments that are passed on to the call to `GET` the (JSON) object, or `method` if specified. For the former, this means that httr and curl options can be given in the call; some issues with timeouts. such as when IPv6 is not properly configured, can be circumvented e.g. if IPv6 is causing trouble, then IPv4 can be forced via the argument `config=config(ipresolve=1))`

Let me know if this is enough, or perhaps I should do something a bit more robust.
